sample.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"os"
+	"strings"
+)
+
+// Sample represents a Google Cloud Platform sample and associated properties.
+type Sample struct {
+	// The Google Cloud Project ID this sample will deploy to
+	googleCloudProject string
+
+	// The local directory this sample is located in
+	dir string
+
+	// The CloudRunService this sample will deploy to
+	cloudRunService *CloudRunService
+
+	// The CloudContainerImage that represents the location of this sample's build container image in the GCP Container
+	// Registry
+	cloudContainerImage *CloudContainerImage
+
+	// The Lifecycle for building and deploying this sample to Cloud Run
+	buildDeployLifecycle *Lifecycle
+}
+
+// newSample creates a new Sample object for the sample located in the provided local directory.
+func newSample(dir string) *Sample {
+	sample := Sample{
+		googleCloudProject: os.Getenv("GOOGLE_CLOUD_PROJECT"),
+		dir:                dir,
+	}
+
+	sample.cloudRunService = newCloudRunService(&sample)
```

Comment:
This pattern, where cloudRunService (and the other objects) both are fields of Sample and have a pointer back to their Sample, is not something that I've seen typically. It feels like things are not encapsulated very well. Is there a way to pass just the data that we need to these objects, rather than the sample itself? It might take some more thinking about what functionality needs to live where.